test/extended/dns/dns_libraries.go

```diff
@@ -0,0 +1,95 @@
+package dns
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+
+	kapiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	watchtools "k8s.io/client-go/tools/watch"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+// checkForPodLogFailures goes through the pod logs and determines if there was a failure
+// by looking for "fail" keyword in the logs.
+func checkForPodLogFailures(f *e2e.Framework, pod *kapiv1.Pod) {
+	By("submitting the pod to kubernetes")
+	podClient := f.ClientSet.CoreV1().Pods(f.Namespace.Name)
+	defer func() {
+		By("deleting the pod")
+		defer GinkgoRecover()
+		podClient.Delete(context.Background(), pod.Name, *metav1.NewDeleteOptions(0))
+	}()
+	updated, err := podClient.Create(context.Background(), pod, metav1.CreateOptions{})
+	if err != nil {
+		e2e.Failf("Failed to create %s pod: %v", pod.Name, err)
+	}
+
+	w, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Watch(context.Background(), metav1.SingleObject(metav1.ObjectMeta{Name: pod.Name, ResourceVersion: updated.ResourceVersion}))
+	if err != nil {
+		e2e.Failf("Failed to watch pods: %v", err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), e2e.PodStartTimeout)
+	defer cancel()
+	if _, err = watchtools.UntilWithoutRetry(ctx, w, PodSucceeded); err != nil {
+		e2e.Failf("Failed: %v", err)
+	}
+
+	By("retrieving the pod logs")
+	r, err := podClient.GetLogs(pod.Name, &kapiv1.PodLogOptions{Container: "querier"}).Stream(context.Background())
+	if err != nil {
+		e2e.Failf("Failed to get pod logs %s: %v", pod.Name, err)
+	}
+
+	scan := bufio.NewScanner(r)
+	for scan.Scan() {
+		line := scan.Text()
+		if strings.Contains(line, "fail") {
+			e2e.Failf("DNS resolution failed")
+		}
+	}
+}
+
+var _ = Describe("[sig-network-edge] DNS lookup", func() {
+	f := e2e.NewDefaultFramework("dns-libraries")
+	// using www.redhat.com just as a sample host for dns queries
+	host := "www.redhat.com"
+
+	// performs a DNS query to make sure Go's DNS resolver works fine with OpenShift DNS.
+	It("using Go's DNS resolver", func() {
+		dnsService, err := f.ClientSet.CoreV1().Services("openshift-dns").Get(context.Background(), "dns-default", metav1.GetOptions{})
+		if err != nil {
+			e2e.Failf("Failed to get default DNS service: %v", err)
+		}
+		r := &net.Resolver{
+			PreferGo: true,
+			Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+				d := net.Dialer{
+					Timeout: time.Millisecond * time.Duration(10000),
+				}
+				return d.DialContext(ctx, network, fmt.Sprintf("%s:%s", dnsService.Spec.ClusterIP, "53"))
+			},
+		}
+		_, err = r.LookupHost(context.Background(), host)
```

Comment:
Isn't this code executing on the test runner?  I think you need to spawn a pod that runs inside the cluster in order to test with the service's cluster IP address.

Reply:
Done.